pkg/controller/jobs/deployment/replicaset/rs_controller.go

```diff
@@ -0,0 +1,208 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replicaset
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta2"
+	"sigs.k8s.io/kueue/pkg/controller/core/indexer"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	"sigs.k8s.io/kueue/pkg/features"
+	"sigs.k8s.io/kueue/pkg/podset"
+	"sigs.k8s.io/kueue/pkg/workloadslicing"
+)
+
+var (
+	// GroupVersionKind identifies the Kubernetes API group, version, and kind
+	// for an apps/v1 ReplicaSet. This is used by the Kueue job framework to
+	// recognize and handle ReplicaSet objects.
+	GroupVersionKind = appsv1.SchemeGroupVersion.WithKind("ReplicaSet")
+
+	// FrameworkName is the registration key for this integration within the
+	// Kueue job framework. It is used when registering ReplicaSet-specific
+	// callbacks, reconcilers, and webhooks.
+	FrameworkName = "replicaset"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:  SetupIndexes,
+		NewJob:        NewJob,
+		NewReconciler: NewReconciler,
+		SetupWebhook:  SetupWebhook,
+		JobType:       &appsv1.ReplicaSet{},
+	}))
+}
+
+// +kubebuilder:rbac:groups=apps,resources=replicasets,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=apps,resources=replicasets/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=replicasets/finalizers,verbs=get;update;patch
+
+// ReplicaSet adapts apps/v1 ReplicaSet to Kueue's jobframework.GenericJob and TopLevelJob.
+type ReplicaSet struct {
+	*appsv1.ReplicaSet
+	suspended bool
+}
+
+// Interface conformance compile time checks.
+var (
+	_ jobframework.GenericJob        = (*ReplicaSet)(nil)
+	_ jobframework.TopLevelJob       = (*ReplicaSet)(nil)
+	_ jobframework.JobWithCustomStop = (*ReplicaSet)(nil)
+)
+
+// NewJob returns a new empty ReplicaSet GenericJob wrapper.
+func NewJob() jobframework.GenericJob {
+	return &ReplicaSet{
+		ReplicaSet: &appsv1.ReplicaSet{},
+	}
+}
+
+// NewReconciler constructs a controller-runtime reconciler for ReplicaSet objects.
+var NewReconciler = jobframework.NewGenericReconcilerFactory(NewJob)
+
+// Object returns the underlying controller-runtime client.Object.
+func (r *ReplicaSet) Object() client.Object { return r.ReplicaSet }
+
+// IsSuspended always returns false since ReplicaSet cannot be suspended.
+func (r *ReplicaSet) IsSuspended() bool {
+	return r.suspended
```

Comment:
Instead of using in-memory I'm wondering we could implement this similarly as for PodGroups - by checking if there exist any Pods which are suspended.

The alternative is to use a scheduling gate "admission" and add it to the ReplicaSet when it is suspended.